Allow casting mut array ref to mut ptr

osa1 opened the pull request:
Allow casting mut array ref to mut ptr

We now allow two new casts:

- mut array reference to mut ptr. Example:

      let mut x: [usize; 2] = [0, 0];
      let p = &mut x as *mut usize;

  We allow casting const array references to const pointers so not
  allowing mut references to mut pointers was inconsistent.

- mut array reference to const ptr. Example:

      let mut x: [usize; 2] = [0, 0];
      let p = &mut x as *const usize;

  This was similarly inconsistent as we allow casting mut references to
  const pointers.

Existing test 'vector-cast-weirdness' updated to test both cases.

Fixes #24151

rust-highfive commented:
r? @lcnr

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1785966541) [(plain)](https://github.com/rust-lang/rust/commit/0d4374484884e9c0177b27363dbcfa4c2bb68728/checks/1785966541/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [ui] ui/array-slice-vec/vector-cast-weirdness.rs stdout ----

error: ui test compiled successfully!
status: exit code: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/array-slice-vec/vector-cast-weirdness.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/array-slice-vec/vector-cast-weirdness" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/array-slice-vec/vector-cast-weirdness/auxiliary"
------------------------------------------

------------------------------------------
stderr:
---
test result: FAILED. 11216 passed; 1 failed; 87 ignored; 0 measured; 0 filtered out; finished in 119.51s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:13:32
```

</details>

lcnr commented on src/test/ui/typeck/issue-24151.rs:
```diff
@@ -0,0 +1,9 @@
+    unsafe {
+        let my_num: &mut [i32; 2] = &mut [10, 20];
+        let my_num: *mut i32 = my_num as *mut i32;
+        println!("{}", *my_num.offset(1));
```
can you move this test to `ui/cast` instead and convert it to `run-pass`?
```suggestion
        assert_eq!(*my_num.offset(1), 20);
```

osa1 replied:
Thanks for the suggestion. Done.

lcnr reviewed (approved):
LGTM

we don't have tests for `&[T; N] as *mut T` and `&mut [T; N] as *const T`afaik. Both of these should error. Can you add them as well?

lcnr commented:
Nominating for `T-lang`, this allows the cast `&mut [T; N] as *mut T` copying the already existing one cast `&[T;N]  as *const T`.

Don't think this is controversial but still needs some `T-lang` input.

osa1 commented:
Thanks for the quick review.

> we don't have tests for &[T; N] as *mut T and &mut [T; N] as *const Tafaik. Both of these should error. Can you add them as well?

Should I do it in the same test file, or in a different file?

osa1 commented:
Shouldn't `&mut [T; N] as *const T` be OK? In my changes I allow this cast currently.

lcnr commented:
> Should I do it in the same test file, or in a different file?

:shrug: don't care, whatever you prefer

> Shouldn't `&mut [T; N] as *const T` be OK? In my changes I allow this cast currently.

we do allow it for `&mut T as *const T` so imo it's fine. Afaict your changes don't allow this though?
You're using `if m_expr.mutbl == m_cast.mutbl` :thinking: Am I misunderstanding something here

osa1 commented:
>  You're using if m_expr.mutbl == m_cast.mutbl thinking Am I misunderstanding something here

Ah, sorry, you're right. I'm confused..

I think we could allow it. I'll wait for the team's feedback.

osa1 commented:
@lcnr just out of curiosity, when is this going to be discussed?

lcnr commented:
the lang team meeting takes place every tuesday, so in 3 days

osa1 commented:
Any news from the meeting, @lcnr ?

osa1 commented:
I found the meeting notes: https://hackmd.io/_-pJa1UxTE6qO1JolgNVRQ I guess there wasn't time to discuss this PR.

nikomatsakis commented:
We discussed this in the [meeting today](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-02-09.md) and decided that we approve of this PR. =) Merge away!

lcnr commented on compiler/rustc_typeck/src/check/cast.rs:
```diff
@@ -767,7 +767,7 @@ impl<'a, 'tcx> CastCheck<'tcx> {
         // array-ptr-cast.
 
-        if m_expr.mutbl == hir::Mutability::Not && m_cast.mutbl == hir::Mutability::Not {
+        if m_expr.mutbl == m_cast.mutbl {
```
do you still want to adjust this to allow for `&mut [T; M] as *const T`? 

I personally don't mind either way:shrug:

osa1 replied:
Given that we allow mut-ref-to-mut-ptr I thought it would be more consistent to allow this as well. I implemented this and updated an existing test to test this case too.

osa1 commented:
Thanks lang team for considering this.

I updated the PR:

- I now allow mut-ref-to-const-ptr casts too
- The existing test `vector-cast-weirdness` updated to test both mut-ref-to-mut-ptr and mut-ref-to-const-ptr cases. Because this test now tests all cases I removed the regression test.
- Rebased the branch.

@lcnr I think this is ready for merging now.

lcnr commented:
@osa1 do we already have a test checking that `&[T; N] as *const T` errors?

if/once that test exists r=me

osa1 commented:
@lcnr 

> do we already have a test checking that &[T; N] as *const T errors?

I'm guessing you meant "&[T; N] as *mut T"? I added tests for that now.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1871199646) [(plain)](https://github.com/rust-lang/rust/commit/2aa1ea4fb59c10a13522fdd45d7e6ce0ec3587bc/checks/1871199646/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
Found 436 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/src/test/ui/array-slice-vec/vector-cast-weirdness.rs:21: line longer than 100 chars
tidy error: /checkout/src/test/ui/array-slice-vec/vector-cast-weirdness.rs:23: line longer than 100 chars
some tidy checks failed

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"
expected success, got: exit code: 1

```

</details>

nikomatsakis commented:
Something I was wondering was whether the language reference, or some other documentation, would need to be updated. I couldn't find an obvious place when looking though. cc @rust-lang/lang-docs

lcnr commented:
@bors r+

bors commented:
:pushpin: Commit d64b749f2cebcfec942ecbbb87e24a9f8cc28469 has been approved by `lcnr`

<!-- @bors r=lcnr d64b749f2cebcfec942ecbbb87e24a9f8cc28469 -->
<!-- homu: {"type":"Approved","sha":"d64b749f2cebcfec942ecbbb87e24a9f8cc28469","approver":"lcnr"} -->

ehuss commented:
I think the table at https://doc.rust-lang.org/nightly/reference/expressions/operator-expr.html#type-cast-expressions should be updated. @osa1 do you think you could open a PR for that?

osa1 commented:
@ehuss sure, will do.

osa1 commented:
Lang ref PR: https://github.com/rust-lang/reference/pull/961

bors commented:
:hourglass: Testing commit d64b749f2cebcfec942ecbbb87e24a9f8cc28469 with merge 663d3127eb8812c66293464cf01bb7a2e5cf3763...
<!-- homu: {"type":"BuildStarted","head_sha":"d64b749f2cebcfec942ecbbb87e24a9f8cc28469","merge_sha":"663d3127eb8812c66293464cf01bb7a2e5cf3763"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1884301389) [(plain)](https://github.com/rust-lang-ci/rust/commit/663d3127eb8812c66293464cf01bb7a2e5cf3763/checks/1884301389/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

lcnr commented:
probably a spurious error

@bors retry

lcnr commented:
it might make sense to mention this in the rel notes, not sure :thinking:

Havvy commented:
It's a change in the language, so yes.

CAD97 commented:
Note: `&mut [T; n] as *const T` is treated as a mutable use (thus getting read/write provenance and invalidating outstanding shared references), whereas `&mut T as *const T` is treated as an immutable use (getting read provenance and leaving outstanding shared references as valid). [Zulip thread to discuss.](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/rustc.201.2E52.20made.20.60.26mut.20.5BT.3B.20N.5D.20as.20*const.20T.60.20a.20valid.20cast)
